### PR TITLE
Convert Unibuild to TS

### PR DESCRIPTION
**Pull request by Slava**
I tried to convert a file to Typescript.

It is still unclear if my local setup actually runs tsc the same way it is supposed to.
Also I took some liberty in naming types and the shapes are probably not 100% correct.

I tried.

### CODE REVIEW

**Comment by Slava** [comment]
@abernix pls messenger bag

File: tools/isobuild/isopack.js [hunk 1]
```diff
@@ -44,215 +44,134 @@ var rejectBadPath = function (p) {
...
+    if (fromFormat === "isopack-1") {
+      // For now, there's no difference in this direction at the isopack level.
+      return convertedData;
     }
   }
 
-  return {metadata, originalVersion};
-});
+  static convertOneStepBackward(data, fromFormat) {
+    var convertedData = _.clone(data);
```
Comment (zionts) [inline tools/isobuild/isopack.js]: Yuck

File: tools/isobuild/isopack.js [hunk 2]
```diff
@@ -44,215 +44,134 @@ var rejectBadPath = function (p) {
...
+    if (fromFormat === "isopack-1") {
```
Comment (zionts) [inline tools/isobuild/isopack.js]: Not to be curdmudgeony, but is there an enum or exported const for this? :L

File: tools/isobuild/isopack.js [hunk 3]
```diff
@@ -1606,12 +1522,91 @@ _.extend(Isopack.prototype, {
...
+}
+
+Isopack.knownFormats = ["unipackage-pre2", "isopack-1", "isopack-2"];
+
+Isopack.convertIsopackFormat = function (data, fromFormat, toFormat) {
+  var fromPos = _.indexOf(Isopack.knownFormats, fromFormat);
+  var toPos = _.indexOf(Isopack.knownFormats, toFormat);
+  var step = fromPos < toPos ? 1 : -1;
+
+  if (fromPos === -1) {
```
Comment (zionts) [inline tools/isobuild/isopack.js]: Nit: aCan we combine this throw with the below?

File: tools/isobuild/unibuild.ts [hunk 4]
```diff
@@ -84,22 +140,28 @@ export class Unibuild {
...
     kind = "main",
     arch,
     unibuildBasePath,
     watchSet,
+  }: {
+    isopack: Isopack,
+    kind: string,
+    arch: string,
+    unibuildBasePath: string,
+    watchSet: WatchSet,
```
Comment (benjamn) [inline tools/isobuild/unibuild.ts]: Shouldn't this return type simply be `Unibuild`?

File: tools/isobuild/unibuild.ts [hunk 5]
```diff
@@ -1,21 +1,77 @@
...
 
+type IsopackConstraint = {
+  package: string;
+  constraint: string;
+  unordered: boolean;
+  weak: boolean;
+};
+
+type PrelinkResource = {
+  type: string;
```
Comment (benjamn) [inline tools/isobuild/unibuild.ts]: This `PrelinkResource` type looks very similar to [the `File` type in `import-scanner.ts`](https://github.com/meteor/meteor/blob/f8ba7a96fe98b441a0f043f2ac5c7e98f56e7c3d/tools/isobuild/import-scanner.ts#L290-L308). I wonder if there's an opportunity to consolidate them?

File: tools/isobuild/isopack.js (hunk 2 above)
Comment (benjamn) [reply tools/isobuild/isopack.js]: This is the kind of thing that will be much easier to improve once everything is TypeScript. Remember that `enum` is a TypeScript-specific feature, and we have only just begun the conversion to TypeScript. I think @Slava made the right call preserving the existing code style in this PR.

**Comment by Slava** [comment]
Sorry to be stalling on this PR, I will find time to address y'all comments

image

**Comment by afrokick** [comment]
@Slava Could you please sign the CLA?

**Comment by filipenevola** [comment]
I'm closing old PRs with problems like tests failing or conflicts. If you are interested in continuing this work please mention me here.

**Pull request closed without merging**